core/src/main/java/fr/sncf/osrd/api/stdcm/graph/STDCMPathfinding.java

```diff
@@ -81,26 +82,36 @@ public static STDCMResult findPath(
 
     /** Make the objective function from the edge locations */
     private static List<TargetsOnEdge<STDCMEdge>> makeObjectiveFunction(
-            Set<Pathfinding.EdgeLocation<SignalingRoute>> endLocations,
-            STDCMGraph graph
+            Set<Pathfinding.EdgeLocation<SignalingRoute>> endLocations
     ) {
         return List.of(edge -> {
             var res = new HashSet<Pathfinding.EdgeLocation<STDCMEdge>>();
-            edge = edge.finishBuildingEdge(graph);
-            if (edge == null)
-                return res;
             for (var loc : endLocations)
                 if (loc.edge().equals(edge.route()))
                     res.add(new Pathfinding.EdgeLocation<>(edge, loc.offset()));
             return res;
         });
     }
 
-    private static double edgeRangeCost(EdgeRange<STDCMEdge> range) {
+    // TODO: remove this function ?
+    // private static double edgeRangeCost(EdgeRange<STDCMEdge> range) {
+    //     var envelope = range.edge().envelope();
+    //     var timeStart = STDCMUtils.interpolateTime(envelope, range.edge().route(), range.start(), 0);
+    //     var timeEnd = STDCMUtils.interpolateTime(envelope, range.edge().route(), range.end(), 0);
+    //     return timeEnd - timeStart + range.edge().addedDelay();
+    // }
+
+    /** Compute the total distance of a path (in s) to an edge location
+     * This estimation of the total distance is used to compare paths in the pathfinding algorithm.
+     * We compute the moment when the train reaches the edge location and 
+     * subtract the totalDepartureTimeShift to compare the paths depending on their duration.
+     * For 2 paths with the same duration, we select the earliest one.*/
+    private static double totalDistanceUntilEdgeLocation(Pathfinding.EdgeLocation<STDCMEdge> range) {
         var envelope = range.edge().envelope();
-        var timeStart = STDCMUtils.interpolateTime(envelope, range.edge().route(), range.start(), 0);
-        var timeEnd = STDCMUtils.interpolateTime(envelope, range.edge().route(), range.end(), 0);
-        return timeEnd - timeStart + range.edge().addedDelay();
+        var timeEnd = STDCMUtils.interpolateTime(
+                envelope, range.edge().route(), range.offset(), range.edge().timeStart()
+        );
+        return Math.max(0, timeEnd - range.edge().totalDepartureTimeShift());
```

Comment:
Why is this `max` call needed? A negative value here would be a bug, an `assert result >= 0` would avoid troubles if it actually happens.